### PR TITLE
import poi way and relation

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

File: src/bin/osm2mimir.rs [hunk 1]
```diff
@@ -416,32 +416,75 @@ impl PoiMatcher {
...
+        (obj.is_node() || obj.is_way() || obj.is_relation()) 
+        && self.poi_types.iter().any(|(poi_tag, poi_types)| {
             obj.tags().get(poi_tag).map_or(false, |poi_type| {
                 poi_types.contains(poi_type)
             })
         })
     }
 }
 
+fn format_id(id: i64) -> String {
```
Comment (datanel) [inline src/bin/osm2mimir.rs]: format_poi_id

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -416,32 +416,75 @@ impl PoiMatcher {
...
+                name: name.to_string(),
+                label: format_label(&admins, city_level, name),
+                coord: coord.clone(),
+                administrative_regions: admins_geofinder.get(&coord),
+                weight: 1,
+            }
+	    },
+	    osmpbfreader::OsmObj::Relation(ref relation) => {
+	        let name = relation.tags.get("name").map_or("", |name| name);
+	        let coord = match mimirsbrunn::boundaries::build_boundary(&relation, &obj_map) {
```
Comment (datanel) [inline src/bin/osm2mimir.rs]: you may create a function `fn get_relation_coord``like `fn get_way_coord`

**Comment by datanel** [comment]
anyone else can approve please? @TeXitoi @antoine-de @xlqian ?

File: src/bin/osm2mimir.rs [hunk 3]
```diff
@@ -218,48 +218,58 @@ fn make_admin_geofinder(admins: &AdminsVec) -> AdminGeoFinder {
     geofinder
 }
 
+fn get_relation_coord(obj_map: &BTreeMap<osmpbfreader::OsmId, osmpbfreader::OsmObj>,
+                      relation: &osmpbfreader::objects::Relation)
+                      -> mimir::Coord {
+    let boundary = mimirsbrunn::boundaries::build_boundary(&relation, &obj_map);
+    boundary.as_ref()
```
Comment (pbougue) [inline src/bin/osm2mimir.rs]: this should more be in boundaries

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -416,61 +428,108 @@ impl PoiMatcher {
...
+            mimir::Poi {
+                id: format_poi_id(relation.id),
+                name: name.to_string(),
+                label: format_label(&adms, city_level, name),
+                coord: coord.clone(),
+                zip_codes: get_zip_codes_from_admins(&adms),
+                administrative_regions: adms,
+                weight: 1,
+            }
+        }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
let coord = match *osmobj {
    osmpbfreader::OsmObj::Node(ref node) => mimir::Coord::new(node.lat, node.lon),
    osmpbfreader::OsmObj::Way(ref way) => get_way_coord(obj_map, way),
    osmpbfreader::OsmObj::Relation(ref relation) => get_relation_coord(obj_map, relation)
};
// and only one creation of `mimir::Poi`
```

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -187,20 +187,15 @@ fn administrative_regions(pbf: &mut OsmPbfReader, levels: BTreeSet<u32>) -> Admi
                 .or_else(|| relation.tags.get("postal_code"))
                 .map(|val| &val[..])
                 .unwrap_or("");
-
-            let boundary = mimirsbrunn::boundaries::build_boundary(&relation, &objects);
-            let coord = coord_centre.or_else(|| {
-                    boundary.as_ref().and_then(|b| b.centroid().map(|c| mimir::Coord(c.0)))
-                })
-                .unwrap_or(mimir::Coord::new(0., 0.));
+			let boundary = build_boundary(&relation, &objects);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: indentation.

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -187,20 +187,15 @@ fn administrative_regions(pbf: &mut OsmPbfReader, levels: BTreeSet<u32>) -> Admi
...
             let admin = mimir::Admin {
                 id: admin_id,
                 insee: insee_id.to_string(),
                 level: level,
                 label: name.to_string(),
                 zip_codes: zip_code.split(';').map(|s| s.to_string()).collect(),
                 weight: Cell::new(0),
-                coord: coord,
+                coord: coord_centre.unwrap_or(make_centroid(&boundary)),
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: you can use `unwrap_or_else` to only compute the centroid if needed.

File: src/bin/osm2mimir.rs [hunk 7]
```diff
@@ -445,48 +419,22 @@ fn parse_poi(osmobj: &osmpbfreader::OsmObj,
...
-                coord: coord.clone(),
-                zip_codes: get_zip_codes_from_admins(&adms),
-                administrative_regions: adms,
-                weight: 1,
-            }
-        }
+    let (id, coord) = match *osmobj {
+        osmpbfreader::OsmObj::Node(ref node) => (node.id, mimir::Coord::new(node.lat, node.lon)),
+        osmpbfreader::OsmObj::Way(ref way) => (way.id, get_way_coord(obj_map, way)),
+        osmpbfreader::OsmObj::Relation(ref relation) => (relation.id, make_centroid(&build_boundary(&relation, &obj_map)))
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: I'm afraid the id is not robust: id in osm are, I think, by type. Thus, Node(42), Way(42) and Rel(42) can co-exists.

The easy solution is `poi:osm:way:42`, and I think we were doing that in ed.

File: src/bin/osm2mimir.rs [hunk 8]
```diff
@@ -445,48 +419,22 @@ fn parse_poi(osmobj: &osmpbfreader::OsmObj,
...
+    };
+
+    let name = osmobj.tags().get("name").map_or("", |name| name);
+    let adms = admins_geofinder.get(&coord);
+    mimir::Poi {
+        id: format_poi_id(id),
+        name: name.to_string(),
+        label: format_label(&adms, city_level, name),
+        coord: coord.clone(),
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: I think this `clone()` is not needed.

File: src/bin/osm2mimir.rs [hunk 9]
```diff
@@ -420,18 +420,20 @@ fn parse_poi(osmobj: &osmpbfreader::OsmObj,
              -> mimir::Poi {
     let admins_geofinder = make_admin_geofinder(admins);
     let (id, coord) = match *osmobj {
-        osmpbfreader::OsmObj::Node(ref node) => (node.id, mimir::Coord::new(node.lat, node.lon)),
-        osmpbfreader::OsmObj::Way(ref way) => (way.id, get_way_coord(obj_map, way)),
-        osmpbfreader::OsmObj::Relation(ref relation) => (relation.id, make_centroid(&build_boundary(&relation, &obj_map)))
+        osmpbfreader::OsmObj::Node(ref node) => (format_poi_id("Node", node.id), mimir::Coord::new(node.lat, node.lon)),
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: I'd prefer all minuscule (ie node, way, relation) to match case of id and osm.

**Comment by datanel** [comment]
tests failed

File: libs/bragi/src/api.rs [hunk 10]
```diff
@@ -146,6 +148,14 @@ impl ApiEndPoint {
                             None => "",
                         }
                         .to_string();
+                    let from = match params.find("from") {
+                            Some(val) => val.as_u64().unwrap_or(0),
+                            None => 0,			
+                        };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]:
```rust
let from = params.find("from").and_then(|val| val.as_u64()).unwrap_or(0);
```

File: libs/bragi/src/api.rs [hunk 11]
```diff
@@ -146,6 +148,14 @@ impl ApiEndPoint {
...
+                    let size = match params.find("size") {
+                            Some(val) => val.as_u64().unwrap_or(0),
+                            None => 0,			
+                        };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: Same

Comment (TeXitoi) [inline libs/bragi/src/api.rs]: Default to something else than 0?

File: libs/bragi/src/api.rs [hunk 12]
```diff
@@ -200,6 +212,14 @@ impl ApiEndPoint {
                 let cnx = self.es_cnx_string.clone();
                 endpoint.handle(move |client, params| {
                     let q = params.find("q").unwrap().as_str().unwrap().to_string();
+                    let from = match params.find("from") {
+                            Some(val) => val.as_u64().unwrap_or(0),
+                            None => 0,			
+                        };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: same

File: libs/bragi/src/api.rs [hunk 13]
```diff
@@ -200,6 +212,14 @@ impl ApiEndPoint {
...
+                    let size = match params.find("size") {
+                            Some(val) => val.as_u64().unwrap_or(0),
+                            None => 0,			
+                        };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: same

File: libs/bragi/src/query.rs [hunk 14]
```diff
@@ -148,16 +148,20 @@ fn build_query(q: &str,
...
+         from: u64,
+         size: u64,
          coord: &Option<model::Coord>,
          shape: Option<Vec<rs_es::units::Location>>)
          -> Result<Vec<mimir::Place>, rs_es::error::EsError> {
     let query = build_query(q, match_type, coord, shape);
 
     let mut client = build_rs_client(&cnx.to_string());
-
+
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: bad spaces

File: libs/bragi/src/api.rs [hunk 15]
```diff
@@ -149,12 +151,12 @@ impl ApiEndPoint {
                         }
                         .to_string();
                     let from = match params.find("from") {
-                            Some(val) => val.as_u64().unwrap_or(0),
-                            None => 0,			
+                            Some(val) => val.as_u64().unwrap_or(DEFAULT_FROM),
+                            None => DEFAULT_FROM,			
                         };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]:
```rust
let from = params.find("from").and_then(|val| val.as_u64()).unwrap_or(DEFAULT_FROM);
```

Same after.

File: libs/bragi/src/query.rs (hunk 14 above)
Comment (azime) [inline libs/bragi/src/query.rs]: ok

File: libs/bragi/src/api.rs (hunk 15 above)
Comment (azime) [inline libs/bragi/src/api.rs]: ok

File: libs/bragi/src/api.rs [hunk 16]
```diff
@@ -145,19 +145,9 @@ impl ApiEndPoint {
...
-                        .to_string();
-                    let from = match params.find("from") {
-                            Some(val) => val.as_u64().unwrap_or(DEFAULT_FROM),
-                            None => DEFAULT_FROM,			
-                        };
-                    let size = match params.find("size") {
-                            Some(val) => val.as_u64().unwrap_or(DEFAULT_SIZE),
-                            None => DEFAULT_SIZE,			
-                        };
+                    let q = params.find("q").unwrap().as_str().unwrap().to_string();
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]:
```rust
let q = params.find("q").and_then(|val| val.as_str()).unwrap_or("").to_string();
```
?

Comment (datanel) [inline libs/bragi/src/api.rs]: why to_string()?

**Comment by datanel** [comment]
is it ok for tou now @TeXitoi ?

**Pull request merged**